packages/client/internals/Goto.vue

```diff
@@ -1,74 +1,189 @@
+  if (selectedItem.value >= autocomplete.value.length)
+    selectedItem.value = -1
+  scroll()
+}
+
+function focusUp(event: Event) {
+  if (autocomplete.value.length > 0)
+    event.preventDefault()
+  selectedItem.value--
+  if (selectedItem.value <= -2)
+    selectedItem.value = autocomplete.value.length - 1
+  scroll()
+}
+
+function scroll() {
+  const item = items.value?.[selectedItem.value]
+  if (item && list.value) {
+    if (item.offsetTop + item.offsetHeight > list.value.offsetHeight + list.value.scrollTop) {
+      list.value.scrollTo({
+        behavior: 'smooth',
+        top: item.offsetTop + item.offsetHeight - list.value.offsetHeight + 1,
+      })
+    }
+    else if (item.offsetTop < list.value.scrollTop) {
+      list.value.scrollTo({
+        behavior: 'smooth',
+        top: item.offsetTop,
+      })
+    }
+  }
+}
+
+function updateText(event: Event) {
+  selectedItem.value = -1
+  text.value = (event.target as HTMLInputElement).value
+}
+
+function select(item: string) {
+  go(+aliases.value.get(item)!)
+  close()
+}
+
 watch(showGotoDialog, async (show) => {
   if (show) {
-    await nextTick()
     text.value = ''
-    input.value?.focus()
+    selectedItem.value = -1
+    // delay the focus to avoid the g character coming from the key that triggered showGotoDialog
+    setTimeout(() => input.value?.focus(), 0)
   }
   else {
     input.value?.blur()
   }
 })
 
-// remove the g character coming from the key that triggered showGotoDialog (e.g. in Firefox)
-watch(text, (t) => {
-  if (t.match(/^[^0-9/]/))
-    text.value = text.value.substring(1)
+watch(activeElement, () => {
+  if (!container.value?.contains(activeElement.value as Node))
+    close()
 })
 </script>
 
 <template>
   <div
     id="slidev-goto-dialog"
-    class="fixed right-5 bg-main transform transition-all"
+    ref="container"
+    class="fixed right-5 transition-all"
     :class="showGotoDialog ? 'top-5' : '-top-20'"
-    shadow="~"
-    p="x-4 y-2"
-    border="~ transparent rounded dark:gray-400 dark:opacity-25"
   >
-    <input
-      ref="input"
-      v-model="text"
-      type="text"
-      :disabled="!showGotoDialog"
-      class="outline-none bg-transparent"
-      placeholder="Goto..."
-      :class="{ 'text-red-400': !valid && text }"
-      @keydown.enter="goTo"
-      @keydown.escape="close"
-      @blur="close"
+    <div
+      class="bg-main transform"
+      shadow="~"
+      p="x-4 y-2"
+      border="~ transparent rounded dark:gray-400 dark:opacity-25"
     >
+      <input
+        ref="input"
+        :value="text"
+        type="text"
+        :disabled="!showGotoDialog"
+        class="outline-none bg-transparent"
+        placeholder="Goto..."
+        :class="{ 'text-red-400': !valid && text }"
+        @keydown.enter="goTo"
+        @keydown.escape="close"
+        @keydown.down="focusDown"
+        @keydown.up="focusUp"
+        @input="updateText"
+      >
+    </div>
+    <ul
+      v-if="autocomplete.length > 0"
```

Comment:
Having auto-complete is a good idea. And since we are inferring the title for each slide, maybe we could display them and even do a fuzzy search based on the title as well?

<img width="223" alt="image" src="https://user-images.githubusercontent.com/11247099/227497293-8e5e9972-b09c-4a92-9c27-d569ea79a626.png">

Reply:
That's what I was proposing in my original message ;)
I agree that it would be nice.
I'm just wondering if this is still relevant for long presentation with a lot of slides.
And in that case maybe we need to add something to distinguish title levels as well in the autocomplete.
Will check.